### PR TITLE
Add Razor TagHelper Parsing Benchmarks

### PR DESCRIPTION
**Pull request by TanayParikh**
Fixes: https://github.com/dotnet/aspnetcore/issues/23454

Based on the ETL Traces from:
https://github.com/dotnet/aspnetcore/issues/23108
https://github.com/dotnet/aspnetcore/issues/23095

This issue turned out to already have been resolved from @ToddGrun's [PR](https://github.com/dotnet/aspnetcore/pull/22887) which brought major performance improvements in this specific area (6.5% less CPU time and 89% less memory allocation).

Baseline before Todd's changes:

|     Mean |   Error |  StdDev |    Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------:|--------:|--------:|--------:|-------:|------:|------:|----------:|
| 100.5 us | 0.58 us | 0.52 us | 9,946.0 | 0.1221 |     - |     - |  27.48 KB |

After Todd's Changes:
|     Mean |    Error |   StdDev |     Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| 94.28 us | 0.897 us | 0.795 us | 10,607.0 |     - |     - |     - |   3.01 KB |

With No Sub-stringing At All:
|     Mean |    Error |   StdDev |     Op/s | Gen 0 | Gen 1 | Gen 2 | Allocated |
|---------:|---------:|---------:|---------:|------:|------:|------:|----------:|
| 93.12 us | 0.222 us | 0.185 us | 10,739.0 |     - |     - |     - |   1.02 KB |

#### **Which proves sub-stringing is not an issue.** 

Added some benchmarks for taghelper parsing so it can help us in potential future investigations.

Benchmarks:
|                                  Method |        Mean |     Error |    StdDev |     Op/s | Allocated |
|---------------------------------------- |------------:|----------:|----------:|---------:|----------:|
|      'TagHelper Design Time Processing' | 2,331.51 us | 28.916 us | 27.048 us |    428.9 | 985.33 KB |
| 'TagHelper Component Directive Parsing' |    90.46 us |  0.472 us |  0.394 us | 11,055.1 |   3.01 KB |

Notes / Attributions:
- `BlazorServerTagHelpers` is just a demo file concatonated from the basic `BlazorServer` files
- `RazorDiagnosticJsonConverter` is copied from: https://github.com/dotnet/aspnetcore/blob/b7d9e8cfea97c9a3e6df871666d489556e2a3d0b/src/Shared/RazorShared/RazorDiagnosticJsonConverter.cs
- `TagHelperDescriptorJsonConverter` is copied from: https://github.com/dotnet/aspnetcore-tooling/blob/73c96f1c00b90e0f9d4fdbfb0905376229ceb913/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs
- `taghelpers.json` was updated from: https://github.com/dotnet/aspnetcore-tooling/blob/73c96f1c00b90e0f9d4fdbfb0905376229ceb913/src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Performance/taghelpers.json
- `ReadTagHelpers` was copied over from https://github.com/dotnet/aspnetcore-tooling/blob/fef50ba62387cb87d34970b4c88290c233361927/src/Razor/benchmarks/Microsoft.AspNetCore.Razor.Performance/ProjectSystem/ProjectSnapshotManagerBenchmarkBase.cs#L83-L93

Fixes: https://github.com/dotnet/aspnetcore/issues/23454

### CODE REVIEW

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperBinderPhase.cs [hunk 1]
```diff
@@ -247,8 +247,8 @@ public ComponentDirectiveVisitor(string filePath, IReadOnlyList<TagHelperDescrip
                     {
                         // If this is a child content tag helper, we want to add it if it's original type is in scope.
                         // E.g, if the type name is `Test.MyComponent.ChildContent`, we want to add it if `Test.MyComponent` is in scope.
-                        TrySplitNamespaceAndType(typeName, out var typeNameTextSpan, out var _);
-                        typeName = GetTextSpanContent(typeNameTextSpan, typeName);
+                        TrySplitNamespaceAndType(typeName, out var namespaceTextSpan, out var _);
```
Comment (ajaybhargavb) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperBinderPhase.cs:250]: Are these just renames?

**Review by ajaybhargavb — APPROVED** [review]
Nice!

> Which proves sub-stringing is not an issue.

The memory allocation went from 3KB to 1KB which is still good.

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperBinderPhase.cs (hunk 1 above)
Comment (TanayParikh) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorTagHelperBinderPhase.cs:250]: Yep to match the signature https://github.com/dotnet/aspnetcore/pull/23627/files/5ed18a194f7d0ddd70b14d9deb1c6f8fc76c97c5#diff-e1c88de1d8233078ce6bb0bc46a2aab7R418

**Comment by TanayParikh** [comment]
> The memory allocation went from 3KB to 1KB which is still good.

To clarify that 3KB -> 1KB was a hypothetical if we weren't to do substringing at all. We've kept the substringing in place as we need to be able to extract the namespace from a fully qualified name. 

The point I was trying to make was that this area has already been optimized substantially (from 27.48KB to 3KB) so any further optimization will have diminishing returns. Note; the ETL's which instigated the issue were from prior to this change.

File: src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/RazorDiagnosticJsonConverter.cs [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+using System;
+using System.Globalization;
+using Microsoft.AspNetCore.Razor.Language;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor.Serialization
+{
+    internal class RazorDiagnosticJsonConverter : JsonConverter
```
Comment (NTaylorMullen) [inline src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/RazorDiagnosticJsonConverter.cs:12]: You shouldn't need to replicate this. We have a converter in the CodeAnalysis.Workspaces dll already.

File: src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/TagHelperDescriptorJsonConverter.cs [hunk 3]
```diff
@@ -0,0 +1,892 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNetCore.Razor.Language;
+using Newtonsoft.Json;
+
+namespace Microsoft.CodeAnalysis.Razor.Serialization
+{
+    internal class TagHelperDescriptorJsonConverter : JsonConverter
```
Comment (NTaylorMullen) [inline src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/TagHelperDescriptorJsonConverter.cs:12]: Ditto

**Review by NTaylorMullen — CHANGES_REQUESTED** [review]

**Review by ryanbrandenburg — APPROVED** [review]

File: src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/RazorDiagnosticJsonConverter.cs (hunk 2 above)
Comment (TanayParikh) [reply src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Serialization/RazorDiagnosticJsonConverter.cs]: Removed, and updated the Performance.csproj with the file compilation references as we do elsewhere

File: src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Microsoft.AspNetCore.Razor.Performance.csproj [hunk 4]
```diff
@@ -16,6 +16,15 @@
 
   <ItemGroup>
     <Compile Include="$(SharedSourceRoot)BenchmarkRunner\*.cs" />
+    <None Include="MSN.cshtml" CopyToOutputDirectory="PreserveNewest" />
+    <None Include="taghelpers.json" CopyToOutputDirectory="PreserveNewest" />
+    <None Include="BlazorServerTagHelpers.razor" CopyToOutputDirectory="PreserveNewest" />
+    <Compile Include="$(SharedSourceRoot)RazorShared\TagHelperDescriptorJsonConverter.cs">
```
Comment (TanayParikh) [inline src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Microsoft.AspNetCore.Razor.Performance.csproj:22]: As done elsewhere:
https://github.com/dotnet/aspnetcore/blob/master/src/Razor/Microsoft.AspNetCore.Razor.Tools/src/Microsoft.AspNetCore.Razor.Tools.csproj#L30-L35

Comment (TanayParikh) [reply src/Razor/perf/Microsoft.AspNetCore.Razor.Performance/Microsoft.AspNetCore.Razor.Performance.csproj:22]: Also, note https://github.com/dotnet/aspnetcore/blob/master/src/Shared/RazorShared/TagHelperDescriptorJsonConverter.cs

is not updated to reflect the most recent: https://github.com/dotnet/aspnetcore-tooling/blob/73c96f1c00b90e0f9d4fdbfb0905376229ceb913/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Serialization/TagHelperDescriptorJsonConverter.cs

Is this update handled automatically, or do we have to do some transfer?

**Review by NTaylorMullen — APPROVED** [review]

**Pull request merged**